lib/dashboard/adapter.go

```diff
@@ -0,0 +1,193 @@
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*timeout)
+	defer cancel()
+
+	// Upload an object with storage.Writer.
+	wc := client.Bucket(bucket).Object(object).NewWriter(ctx)
+	if _, err = io.Copy(wc, f); err != nil {
+		return fmt.Errorf("io.Copy: %v", err)
+	}
+	if err := wc.Close(); err != nil {
+		return fmt.Errorf("Writer.Close: %v", err)
+	}
+	return nil
+}
+
+// GetAllVulnerabilities gets all of the vulnerability occurrences associated
+// with images in a specific project using the Container Analysis Service.
+// nolint[errcheck]
+func GetAllVulnerabilities(
+	projectID string,
+) ([]*grafeaspb.Occurrence, error) {
+	ctx := context.Background()
+	client, err := containeranalysis.NewClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("NewClient: %v", err)
+	}
+	defer client.Close()
+
+	req := &grafeaspb.ListOccurrencesRequest{
+		Parent: fmt.Sprintf("projects/%s", projectID),
+		Filter: fmt.Sprintf("kind = %q", "VULNERABILITY"),
+	}
+
+	var occurrenceList []*grafeaspb.Occurrence
+	it := client.GetGrafeasClient().ListOccurrences(ctx, req)
+	for {
+		occ, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("occurrence iteration error: %v", err)
+		}
+		occurrenceList = append(occurrenceList, occ)
+	}
+	return occurrenceList, err
+}
+
+func parseImageResourceURL(resourceURL string) (string, string) {
+	FQIN := path.Base(resourceURL)
+	splitFQIN := strings.Split(FQIN, "@")
+	return splitFQIN[0], splitFQIN[1]
+}
+
+func parseVulnName(noteName string) string {
+	return path.Base(noteName)
+}
+
+// GenerateDashboardJSON parses the a slice of vulnerability occurrences
+// into a JSON that only contains the necessary information for each
+// vulnerability. The JSON is also formatted to be easily processed into
+// an HTML table by the vulnerability dashboard.
+func GenerateDashboardJSON(
+	productionVulnerabilities []*grafeaspb.Occurrence,
+) ([]byte, error) {
+	var occByResource = make(map[string]ImageVulnOcc)
+	var resourceURIs []string
+	for _, occ := range productionVulnerabilities {
+		_, found := occByResource[occ.ResourceUri]
+		if !found {
+			imageName, imageDigest := parseImageResourceURL(occ.ResourceUri)
+			resourceURIs = append(resourceURIs, occ.ResourceUri)
+			occByResource[occ.ResourceUri] = ImageVulnOcc{
+				imageName,
+				imageDigest,
+				0,
+				[]string{},
+				[]string{},
+			}
+		}
+		imageVulnOcc := occByResource[occ.ResourceUri]
+		imageVulnOcc.NumVulnerabilities++
+		vulnName := parseVulnName(occ.NoteName)
+		vuln := occ.GetVulnerability()
+		if vuln.GetSeverity() == grafeaspb.Severity_CRITICAL {
+			imageVulnOcc.CriticalVulnerabilities = append(
+				imageVulnOcc.CriticalVulnerabilities,
+				vulnName,
+			)
+		}
+		if vuln.GetFixAvailable() {
+			imageVulnOcc.FixableVulnerabilities = append(
+				imageVulnOcc.FixableVulnerabilities,
+				vulnName,
+			)
+		}
+		occByResource[occ.ResourceUri] = imageVulnOcc
+	}
+
+	sort.Strings(resourceURIs)
```

Comment:
What does this section do?

Reply:
Maybe add a comment there that says it adds determinism?

Reply:
I think if you use yaml instead of json the sorting will happen automatically. Not sure, please check.

Reply:
https://stackoverflow.com/questions/18668652/how-to-produce-json-with-sorted-keys-in-go

Reply:
yeah the JSON marshaling a map is keeping everything ordered. originally it wasn't working for me because I wanted to marshal a list for purposes of parsing in the JavaScript file; but I've since updated the JavaScript so marshaling a map should work well.